tests/integration/test_upgrade.py

```diff
@@ -0,0 +1,257 @@
+            "db_host": kube_core_client.read_namespaced_pod(
+                name="mysql", namespace=ops_test.model_name
+            ).status.pod_ip,
+            "db_name": "wordpress",
+            "db_user": "wordpress",
+            "db_password": "wordpress-password",
+        }
+        if swift_config:
+            logger.info(f"deploy with swift: {swift_config}")
+            charm_config["wp_plugin_openstack-objectstorage_config"] = json.dumps(swift_config)
+        return charm_config
+
+    return _gen_upgrade_test_charm_config
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_old_version(
+    num_units,
+    ops_test: pytest_operator.plugin.OpsTest,
+    application_name,
+    kube_core_client,
+    deploy_and_wait_for_mysql_pod,
+    gen_upgrade_test_charm_config,
+):
+    """
+    arrange: none.
+    act: deploy all required charms and kubernetes pods for tests.
+    assert: all charms and pods are deployed successfully.
+    """
+    assert ops_test.model
+
+    async def deploy_wordpress():
+        await ops_test.model.deploy(
+            "wordpress-k8s",
+            resources={"wordpress-image": "wordpresscharmers/wordpress:v5.9.4-20.04_edge"},
+            application_name=application_name,
+            num_units=num_units,
+        )
+
+    await asyncio.gather(
+        deploy_wordpress(),
+        deploy_and_wait_for_mysql_pod(),
+        ops_test.run("playwright", "install", "chromium"),
+    )
+    await ops_test.model.applications[application_name].set_config(gen_upgrade_test_charm_config())
+    await ops_test.model.wait_for_idle(status=ops.model.ActiveStatus.name)  # type: ignore
+
+
+@pytest.mark.abort_on_fail
+async def test_create_example_blog(
+    ops_test,
+    default_admin_password,
+    unit_ip_list,
+    application_name,
+    test_image,
+    screenshot_dir,
+    kube_core_client,
+    swift_config,
+):
+    """
+    arrange: all charms and pods are deployed successfully.
+    act: enable some plugins and create posts and upload some images to the test WordPress blog.
+    assert: all operations finished without error.
+    """
+    namespace = ops_test.model_name
+
+    def get_wordpress_podspec_pod():
+        for pod in kube_core_client.list_namespaced_pod(namespace=namespace).items:
+            name = pod.metadata.name
+            if name.startswith("wordpress") and not name.endswith("-0"):
+                return name
+        raise KeyError("can not find wordpress pod")
+
+    wordpress_pod = get_wordpress_podspec_pod()
+
+    def kubernetes_exec(cmd):
+        logger.info(f"exec {cmd} on {wordpress_pod}")
+        resp = kubernetes.stream.stream(
+            kube_core_client.connect_get_namespaced_pod_exec,
+            name=wordpress_pod,
+            namespace=namespace,
+            container="wordpress",
+            command=cmd,
+            stderr=True,
+            stdin=False,
+            stdout=True,
+            tty=False,
+        )
+        logger.info("Response: " + resp)
```

Comment:
Wouldn't it be better to use pebble exec which is more the "Canonical" way ?

Reply:
I'd even keep a bash script, copy it to the charm, and run it instead of doing multiple exec in a row

Reply:
The current CharmHub version of the charm lacks some functionalities, after the CharmHub version of the charm deployed and configured, some manual operations is still needed to make the WordPress blog work. All the `exec` here is to run those manual operations automatically. This is just for testing and it will be removed once we published our version to CharmHub.

Reply:
Ok I get, but wasn't pebble usable instead of kubernetes client for the exec method ?

Reply:
The old version of the charm is downloaded and deployed from CharmHub. So, it's difficult to alter to code of the charm to run some pebble commands. Besides, all commands executed here can actually run everywhere with PHP installed, I am choosing to run inside the WordPress pod container because the PHP is already installed there, it's easier and time-saving to just run some commands there.